test/metabase/driver/sql/query_processor_test.clj

```diff
@@ -465,3 +470,26 @@
                               :alias        "CategoriesStats"
                               :fields       :all}]
                :limit       3}))))))
+
+(deftest expressions-and-coercions-test
+  (mt/test-drivers (conj (sql-jdbc.tu/sql-jdbc-drivers) :bigquery)
+    (testing "Don't cast in both inner select and outer select when expression (#12430)"
+      (mt/with-temp-copy-of-db
+        (let [price-field-id (data/id :venues :price)]
+          (db/update! Field price-field-id {:coercion_strategy :Coercion/UNIXSeconds->DateTime
+                                            :effective_type    :type/DateTime})
+          (let [results (qp/process-query {:database   (data/id)
+                                           :query      {:source-table (data/id :venues)
+                                                        :expressions  {:test ["*" 1 1]}
+                                                        :fields       [[:field price-field-id nil]
+                                                                       [:expression "test"]]}
+                                           :middleware {:format-rows? false}
+                                           :type       "query"})]
+            (is (schema= [(s/one (if (#{:sqlite} driver/*driver*)
+                                   s/Str
+                                   (s/pred #(instance? java.time.temporal.Temporal %)))
+                                 "date")
+                          (s/one (if (#{:oracle} driver/*driver*)
+                                   (s/pred decimal? 'decimal) ;; oracle returns 1M annoyingly
+                                   s/Int) "expression")]
+                         (-> results :data :rows first)))))))))
```

Comment:
1. You should use `mt/rows` instead of `:data :rows`. Otherwise if the query returns an error, `(-> :data :rows)` will return `nil`, which is super confusing. `mt/rows` will only return the rows if they're present, and if there's an error in the query it will fail
2. We have `mt/formatted-rows` for cases when different drivers return the same value as different types. We use `mt/formatted` rows all over the codebase to make sure numbers are integers to deal with the Oracle nonsense
3. `:sqlite` only returns something different than other drivers because you're including the `:middleware {:format-rows? false}` option... that has nothing to do with this test, so why not remove that option so you don't need to special-case `:sqlite`?